src/rules/no-t-call-in-react-component.ts

```diff
@@ -0,0 +1,314 @@
+import { ParserServices, TSESTree } from '@typescript-eslint/utils'
+import { createRule } from '../create-rule'
+import { RuleContext, SourceCode } from '@typescript-eslint/utils/ts-eslint'
+import ts, { type Expression, type SourceFile, type TypeChecker } from 'typescript'
+
+type Node = TSESTree.Node
+const { AST_NODE_TYPES } = TSESTree
+export const name = 'no-t-call-in-react-component'
+export const rule = createRule({
+  name,
+  meta: {
+    docs: {
+      description: 'use _(msg`...`) instead of t`...` and t() in react components',
+      recommended: 'error',
+    },
+    messages: {
+      default:
+        't`...` and t() call should not be used in react components, use _(msg`...`) instead',
+      fix1: 'Replace with _(msg`...`)',
+      fix2: 'Replace with <Trans>...</Trans>',
+    },
+    fixable: 'code',
+    hasSuggestions: true,
+    schema: [
+      {
+        type: 'object',
+        properties: {},
+        additionalProperties: false,
+      },
+    ],
+    type: 'problem' as const,
+  },
+
+  defaultOptions: [],
+
+  create: (context) => {
+    return {
+      'TaggedTemplateExpression[tag.name=t]'(node: TSESTree.TaggedTemplateExpression) {
+        const hookContainer = findAncestor(node, isHookContainer)
+        if (!hookContainer) return
+        const sourceCode = context.sourceCode
+
+        context.report({
+          node,
+          messageId: 'default',
+          suggest: [
+            {
+              messageId: 'fix1',
+              *fix(fixer) {
+                if (hookContainer.body.type !== AST_NODE_TYPES.BlockStatement) return
+                yield fixer.insertTextBefore(
+                  sourceCode.ast.body[0],
+                  "import { useLingui } from '@lingui/react';import { msg } from '@lingui/macro';",
+                )
+                yield fixer.insertTextBefore(
+                  hookContainer.body.body[0],
+                  'const { _ } = useLingui();',
+                )
+                yield fixer.replaceText(node, `_(msg${sourceCode.getText(node.quasi)})`)
+              },
+            },
+            {
+              messageId: 'fix2',
+              *fix(fixer) {
+                if (!(context.filename.endsWith('.tsx') || context.filename.endsWith('.jsx')))
+                  return
+                const parser = getParserServices(context)
+                if (!parser) return
+
+                const fixable = canFixAsJSX(parser, sourceCode, node)
+                if (!fixable) return
+                const template = node.quasi
+                if (
+                  template.type !== AST_NODE_TYPES.TemplateLiteral &&
+                  template.type !== AST_NODE_TYPES.Literal
```

Comment:
There are a little type guard helpers in the `helpers.ts`, they make code a little bit cleaner. 

```js
template.type !== AST_NODE_TYPES.Literal

!isLiteral(template)
```
You can use it and extend with your own, because there is not a full set.